numba_dppy/dpnp_glue/dpnp_randomimpl.py

```diff
@@ -0,0 +1,1287 @@
+            res = np.empty(size, dtype=res_dtype)
+            if res.size != 0:
+                common_impl_1_arg(p, res, dpnp_func, PRINT_DEBUG)
+            return res
+
+    return dpnp_impl
+
+
+@overload(stubs.dpnp.gumbel)
+def dpnp_random_impl(loc=0.0, scale=1.0, size=None):
+    name = "gumbel"
+    dpnp_lowering.ensure_dpnp(name)
+
+    ret_type = types.void
+    """
+    dpnp source:
+    https://github.com/IntelPython/dpnp/blob/0.4.0/dpnp/backend/custom_kernels_random.cpp#L154
+
+    Function declaration:
+    void custom_rng_gumbel_c(void* result, double loc, double scale, size_t size)
+    """
+    sig = signature(ret_type, types.voidptr, types.float64, types.float64, types.intp)
+    dpnp_func = dpnp_ext.dpnp_func("dpnp_" + name, ["float64", "NONE"], sig)
+    res_dtype = np.float64
+    PRINT_DEBUG = dpnp_lowering.DEBUG
+
+    if not isinstance(loc, float):
+        if not (isinstance(loc, types.Float)):
+            raise ValueError("We only support scalar for input: loc")
+
+    if not isinstance(scale, float):
+        if not (isinstance(scale, types.Float)):
+            raise ValueError("We only support scalar for input: scale")
+
+    if size in (None, types.none):
+
+        def dpnp_impl(loc=0.0, scale=1.0, size=None):
+            res = np.empty(1, dtype=res_dtype)
+            common_impl_2_arg(loc, scale, res, dpnp_func, PRINT_DEBUG)
+            return res[0]
+
+    else:
+
+        def dpnp_impl(loc=0.0, scale=1.0, size=None):
+            res = np.empty(size, dtype=res_dtype)
+            if res.size != 0:
+                common_impl_2_arg(loc, scale, res, dpnp_func, PRINT_DEBUG)
+            return res
+
+    return dpnp_impl
+
+
+@overload(stubs.dpnp.hypergeometric)
+def dpnp_random_impl(ngood, nbad, nsample, size=None):
+    name = "hypergeometric"
+    dpnp_lowering.ensure_dpnp(name)
+
+    ret_type = types.void
+    """
+    dpnp source:
+    https://github.com/IntelPython/dpnp/blob/0.4.0/dpnp/backend/custom_kernels_random.cpp#L169
+
+    Function declaration:
+    void custom_rng_hypergeometric_c(void* result, int l, int s, int m, size_t size)
+    """
+    sig = signature(
+        ret_type, types.voidptr, types.int32, types.int32, types.int32, types.intp
+    )
+    dpnp_func = dpnp_ext.dpnp_func("dpnp_" + name, ["int32", "NONE"], sig)
+
+    res_dtype = np.int32
+
+    PRINT_DEBUG = dpnp_lowering.DEBUG
+
+    @register_jitable
+    def common_impl(ngood, nbad, nsample, res):
+        sycl_queue = dpctl_functions.get_current_queue()
+        res_usm = dpctl_functions.malloc_shared(res.size * res.itemsize, sycl_queue)
+
+        dpnp_func(res_usm, ngood, nbad, nsample, res.size)
+
+        dpctl_functions.queue_memcpy(
+            sycl_queue, res.ctypes, res_usm, res.size * res.itemsize
+        )
+
+        dpctl_functions.free_with_queue(res_usm, sycl_queue)
+
+        dpnp_ext._dummy_liveness_func([res.size])
+
+        if PRINT_DEBUG:
+            print("DPNP implementation")
```

Comment:
It will register new function in numba each time this function is called.
Can we move this `register_jitable` outside of the function `dpnp_random_impl`?

Reply:
I'll have to look into it. I think we can.